includes/class-wc-payments.php

```diff
@@ -754,6 +799,41 @@ public static function get_file_version( $file ): string {
 		return WCPAY_VERSION_NUMBER;
 	}
 
+	/**
+	 * Returns payment method instance by Stripe ID.
+	 *
+	 * @param string $payment_method_id Stripe payment method type ID.
+	 * @return UPE_Payment_Method Matching UPE Payment Method instance.
+	 */
+	public static function get_payment_method_by_id( $payment_method_id ) {
+		if ( ! isset( self::$upe_payment_method_map[ $payment_method_id ] ) ) {
+			return false;
```

Comment:
👋  @FangedParakeet @mdmoore, this is falgged by `npm run psalm`.  For the `get_payment_method_by_id()`, we now update it to `return false` if the payment method isn't found.  But there are a few calls to this function that don't handle `false` very well. Can we assume the `self::$upe_payment_method_map[ $payment_method_id ]` will always exist and will return a `UPE_Payment_Method`? 

Or, should all calls to this function assume that the payment method may not exist and the caller will need to handle error properly? ie. skip the payment method, throw exception , etc

Reply:
[This](https://github.com/Automattic/woocommerce-payments/pull/4557/commits/bde68cd09312b4317f8f19e27c61ac517dd26b38) is what I did to address the psalm errors. So in terms of dealing with the psalm's errors, it will be fixed.

I was wondering more around if these functions do return `false`, what's gonna happen to all the code that calls them. Maybe we will need to go add `(if !false && ...` in various places.

Reply:
> Can we assume the self::$upe_payment_method_map[ $payment_method_id ] will always exist and will return a UPE_Payment_Method?

As far as I can see, the `WC_Payments::get_payment_method_by_id()` function is only fed payment method IDs that are returned by `get_upe_enabled_payment_method_ids()` or `get_payment_method_ids_enabled_at_checkout()`, which both basically rely on these values being stored as a meta value in the `options` table. I _think_ the only way to get something invalid in there would be to manually edit those meta values, so that should be a pretty negligible notion. 

I think it's just overly defensive programming, to prevent the situation where some versions of PHP might swallow the undefined offset error and others might throw an exception. Perhaps we could throw our own exception instead, but I don't think this is really a case that can be reasonably encountered, in any case.

Reply:
>Can we assume the self::$upe_payment_method_map[ $payment_method_id ] will always exist and will return a UPE_Payment_Method?
> Or, should all calls to this function assume that the payment method may not exist and the caller will need to handle the error properly? ie. skip the payment method, throw exception, etc

I'm curious about how to handle things like this as well. At the shallow level, does returning `null` make more sense than `false` here since we're trying to get something that may not exist? Because `true/false` implies the question you're asking is binary (does this thing exist?) rather than _return me the thing I'm asking for_. 
Maybe a pedantic question but I'm curious about it as a newer engineer. 🤔  

At a deeper level, I wonder about the suggestion @FangedParakeet made about throwing an Exception if we don't find a payment method. Throwing an exception could make sense if we call the method incorrectly with an illegitimate payment method ID (we want to fail hard if we screw up rather than "hide" that with defensive programming and conditionals). This assumes that we could know beforehand if it was illegitimate.

But it could be a design decision that when we loop over the available payment methods and encounter a payment method that doesn't exist (when `WC_Payments::get_payment_method_by_id()` returns `false`, we simply short-circuit and don't add it to the list of usable payment methods). Seems like it would be nice to have those pruned away to avoid it happening over and over again 🤔 

By the way, it seems possible that we'd accidentally be [calling methods on a non-object here](https://github.com/Automattic/woocommerce-payments/pull/3968/files#diff-b541c0a287f40f82a1d0e43c52c7c7a2b67ae233f261655cbb437726fec83feeR1076) if we return `false`:

```
$payment_method  = WC_Payments::get_payment_method_by_id( $payment_method_id );
	$settings[ $payment_method_id ] = [
	'isReusable'           => $payment_method->is_reusable(),
	'title'                => $payment_method->get_title(),
    ];
```

It's probably all pretty shallow and wouldn't cause us any headache; I think I'm mostly curious from a learning perspective if anyone has thoughts on it.

Reply:
WCPay initializes the `WC_Payments::$upe_payment_method_map` in its `WC_Payments::init()` function. This loops all our supported [payment methods](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/class-wc-payments.php#L312-L325).

@FangedParakeet pointed out in the [earlier comment](https://github.com/Automattic/woocommerce-payments/pull/3968#discussion_r947091062) that the `$payment_method_id` passed to `get_payment_method_ids_enabled_at_checkout()` here is controlled by capability things:
```php
		$capturable_payment_methods = $capture ? $this->get_upe_enabled_payment_method_ids() : [ 'card' ];
		$enabled_payment_methods    = [];
		$active_payment_methods     = $this->get_upe_enabled_payment_method_statuses();
		foreach ( $capturable_payment_methods as $payment_method_id ) {
			$payment_method_capability_key = $this->payment_method_capability_key_map[ $payment_method_id ] ?? 'undefined_capability_key';
			if ( false !== $this->wc_payments_get_payment_method_by_id( $payment_method_id ) ) {

```
And essentially `$payment_method_id` is determined by `get_upe_enabled_payment_methods_ids()` which is from the option table:
```php
	public function get_upe_enabled_payment_method_ids() {
		return $this->get_option(
			'upe_enabled_payment_method_ids',
			[
				'card',
			]
		);
	}
```

So we know that it is either `card` or something from the `options` table. With all these, the only way for `WC_Payments::get_payment_method_by_id( $payment_method_id );` to be `false` is when the option table contain payment method other than the ones we listed in our `WC_Payments::init()`, which seems unlikely. 

Because this case is unexpected, I think I would choose to throw an `Exception` inside `WC_Payments::get_payment_method_by_id()` rather than failing gracefully by skipping over `false` or `null`. 

Also, because the payment method should always be one of the payment methods we created in `WC_Payments::init(), I think we don't need this condition anymore: `if ( false !== $this->wc_payments_get_payment_method_by_id( $payment_method_id ) ) {`